monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/MergePrioritizedListObservable.scala

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) 2014-2020 by The Monix Project Developers.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.reactive.internal.builders
+
+import java.util.concurrent.PriorityBlockingQueue
+
+import monix.execution.Ack.{Continue, Stop}
+import monix.execution.cancelables.CompositeCancelable
+import monix.execution.{Ack, Cancelable, Scheduler}
+import monix.reactive.Observable
+import monix.reactive.observers.Subscriber
+
+import scala.concurrent.{Future, Promise}
+import scala.jdk.CollectionConverters._
+import scala.util.Success
+
+private[reactive] final class MergePrioritizedListObservable[A](sources: Seq[Observable[A]], priorities: Seq[Int])
+  extends Observable[A] {
+  require(sources.size == priorities.size, "sources.size != priorities.size")
+
+  override def unsafeSubscribeFn(out: Subscriber[A]): Cancelable = {
+    import out.scheduler
+
+    val numberOfObservables = sources.size
+
+    val lock = new AnyRef
+    var isDone = false
+
+    // NOTE: We use arrays and other mutable structures here to be as performant as possible.
+
+    // MUST BE synchronized by `lock`
+    var lastAck = Continue: Future[Ack]
+
+    case class PQElem(data: A, promise: Promise[Ack], priority: Int) extends Comparable[PQElem] {
+      override def compareTo(o: PQElem): Int =
+        priority.compareTo(o.priority)
+    }
+
+    val pq = new PriorityBlockingQueue[PQElem](sources.size)
```

Comment:
Could we use normal `PriorityQueue` instead of a blocking one? It seems like the access is synchronized anyway

Reply:
Sure, I will look at replacing it and adding necessary synchronization.

Otherwise it looks good?

Reply:
Sorry, I haven't got a moment to take a close look, but it seems to look good. 

I think it's not too far from `mergeMapPrioritized` version If you'd like to go further:
- `onNext` of the source `Observable` (upstream) would add a new `Observable` to the composite instead of doing it from the starting list
- `RefCountObservable` is used to control the number of active children and their lifecycle
- It would be awesome to support different overflow strategies, but it would require to update `BufferedSubscriber` to allow priority queues. The implementation is in `AbstractBackPressuredBufferedSubscriber` for Back-Pressure strategy. Unfortunately I don't know if there is a good implementation of concurrent priority queue out there. I imagine it's too much work, and I don't have capacity to work on it right now so I suppose we can stay with the current behavior which is a back-pressure with bufferSize of 0.

If you don't feel like doing it - that's okay, `mergePrioritizedList` is already better than the status quo

It might also be nice to add a benchmark vs normal merge, but that's just my curiosity and completely optional.

Reply:
Since updating `BufferedSuscriber` would be lots of work I kept my implementation as is. I swapped out the Java `PriorityBlockingQueue` for the Scala `mutable.PriorityQueue`, and added tests.